src/Servers/Kestrel/Core/src/Internal/Http2/HPack/Http2HPackEncoder.cs

```diff
@@ -0,0 +1,385 @@
+                    }
+
+                    length = currentLength;
+                    return false;
+                }
+
+                currentLength += headerLength;
+            }
+            while (headersEnumerator.MoveNext());
+
+            length = currentLength;
+            return true;
+        }
+
+        private bool EncodeHeader(Span<byte> buffer, int staticTableIndex, string name, string value, out int bytesWritten)
+        {
+            // Never index sensitive value.
+            if (_sensitivityDetector?.IsSensitive(name, value) ?? false)
+            {
+                var index = ResolveDynamicTableIndex(staticTableIndex, name);
+
+                return index == -1
+                    ? HPackEncoder.EncodeLiteralHeaderFieldNeverIndexingNewName(name, value, buffer, out bytesWritten)
+                    : HPackEncoder.EncodeLiteralHeaderFieldNeverIndexing(index, value, buffer, out bytesWritten);
+            }
+
+            // No dynamic table. Only use the static table.
+            if (_maxHeaderTableSize == 0)
+            {
+                return staticTableIndex == -1
+                    ? HPackEncoder.EncodeLiteralHeaderFieldWithoutIndexingNewName(name, value, buffer, out bytesWritten)
+                    : HPackEncoder.EncodeLiteralHeaderFieldWithoutIndexing(staticTableIndex, value, buffer, out bytesWritten);
+            }
+
+            // Header is greater than the maximum table size.
+            // Don't attempt to add dynamic header as all existing dynamic headers will be removed.
+            if (HeaderField.GetLength(name.Length, value.Length) > _maxHeaderTableSize)
+            {
+                var index = ResolveDynamicTableIndex(staticTableIndex, name);
+
+                return index == -1
+                    ? HPackEncoder.EncodeLiteralHeaderFieldWithoutIndexingNewName(name, value, buffer, out bytesWritten)
+                    : HPackEncoder.EncodeLiteralHeaderFieldWithoutIndexing(index, value, buffer, out bytesWritten);
+            }
+
+            return EncodeDynamicHeader(buffer, staticTableIndex, name, value, out bytesWritten);
+        }
+
+        private int ResolveDynamicTableIndex(int staticTableIndex, string name)
+        {
+            if (staticTableIndex != -1)
+            {
+                // Prefer static table index.
+                return staticTableIndex;
+            }
+
+            return CalculateDynamicTableIndex(name);
+        }
+
+        private bool EncodeDynamicHeader(Span<byte> buffer, int staticTableIndex, string name, string value, out int bytesWritten)
+        {
+            var headerField = GetEntry(name, value);
+            if (headerField != null)
+            {
+                // Already exists in dynamic table. Write index.
+                var index = CalculateDynamicTableIndex(headerField.Index);
+                return HPackEncoder.EncodeIndexedHeaderField(index, buffer, out bytesWritten);
```

Comment:
> One way to maximize re-use is to limit churn in the table to avoid evicting commonly re-used values. To do this you need to avoid adding values that will rarely be re-used, especially larger values.

It's probably also be a good idea for EncodeDynamicHeader to move the HPackHeaderEntry to the start of the linked list even when it already exists so you get proper LRU behavior.

Reply:
Does HPack allow re-ordering?

Reply:
No. Reordering on the server would mean the client and server evict headers in a different order and things would fall apart.

Reply:
I see. I thought we had more control over exactly what's evicted when adding new entries.